### DISCUSSION TITLE
where is the ~applyUpdate~ installUpdatesIfAvailable function in v2.4.1?

### DISCUSSION THREAD

**Discussion by mbsanchez** [sparkle-project/Sparkle#2378] [Q&A] [answered]
In Sparkle version ~1.8.0~ **1.20.0** we used to use the **applyUpdate** function, which allowed us to ask the user to install the update when he is ready, giving him the chance to finish some work if needed. Technically, we disable **automaticallyChecksForUpdates** and **automaticallyDownloadsUpdates**, then we used **checkForUpdateInformation** which told us if an update is available, then we showed a floating notification in the app that the user can close or just click, "**install**". At that moment we used the **~applyUpdate~** **installUpdatesIfAvailable** function that started downloading the package, installing it, and restarting the app. 

Now, in version 2.4.1 the function **~applyUpdate~** **installUpdatesIfAvailable** was removed we just have the function **checkForUpdateInformation** but I think that this function is worth nothing without **~applyUpdate~** **installUpdatesIfAvailable**.

I'm just wondering how I can achieve the same behavior as in version 1.8.0 on this new version 2.4.1?

**Comment by zorgiepoo** [comment]
> In Sparkle version 1.8.0 we used to use the applyUpdate function, which allowed us to ask the user to install the update when he is ready, giving him the chance to finish some work if needed

I don't know what this `applyUpdate` is. `1.8.0` is incredibly old (2014); I find it quite unlikely you were still using it. Perhaps you were even using some fork (or you're just not describing the old API well).

Checking for updates, letting the user know they can install an update, and then checking for updates again is inefficient if that was what you were previously doing.

[Gentle update reminders](https://sparkle-project.org/documentation/gentle-reminders/) can be done so that the you can provide a gentle reminder like you're doing now and the user when they're ready can go to see the update's release notes and install the update.

Otherwise a more advanced use case would be to implement a custom user interface (via SPUUserDriver) instead of using Sparkle's standard UI (not as lightweight).

Let Sparkle check for updates for you (re-enable automaticallyChecksForUpdates). If this were a new app I would not recommend using automaticallyChecksForUpdates / automaticallyDownloadsUpdates (read their documentation). But you already have these properties persisted in the user defaults.

**Reply by mbsanchez** [reply]
> I don't know what this applyUpdate is. 1.8.0 is incredibly old (2014); I find it quite unlikely you were still using it. Perhaps you were even using some fork (or you're just not describing the old API well).

Sorry, I've made a mistake it is Sparkle 1.20.0, and I've used it because more recent versions have deprecated DSA signatures. Moreover is not **applyUpdate**, the method I was using is **installUpdatesIfAvailable**. Right now  I'm trying to do things correctly, and I want to update to the latest version of Sparkle as my app is double-signed and I think I can update the signature properly. However, I don't want to allow users to skip versions or the app to show any extra popups.

> Checking for updates, letting the user know they can install an update, and then checking for updates again is inefficient if that was what you were previously doing.

What I was previously doing is to **checkForUpdateInformation**, which just tells me that an update is available, then I inform the user about this showing a floating popup, and when the user clicks "install update" button, I just called **installUpdatesIfAvailable** which installed the update without showing any window for confirmation.

I wanted to implement a custom **SPUUserDriver** but I couldn't find a good example for this, and it seems that I need to implement all the methods declared on SPUUserDriver which could be overwhelming. Do we have a **SPUUserDriverAdapter**? I mean a class that has all the methods implemented and for what I just have to overwrite a specific method like **showReadyToInstallAndRelaunch**. 
Thank you!

**Reply by zorgiepoo** [reply]
> Sorry, I've made a mistake it is Sparkle 1.20.0, and I've used it because more recent versions have deprecated DSA signatures. Moreover is not applyUpdate, the method I was using is installUpdatesIfAvailable. Right now I'm trying to do things correctly, and I want to update to the latest version of Sparkle as my app is double-signed and I think I can update the signature properly. However, I don't want to allow users to skip versions or the app to show any extra popups.

Ok, yes, installUpdatesIfAvailable was removed.

> What I was previously doing is to checkForUpdateInformation, which just tells me that an update is available, then I inform the user about this showing a floating popup, and when the user clicks "install update" button, I just called installUpdatesIfAvailable which installed the update without showing any window for confirmation.

Yes, both checkForUpdateInformation and installUpdatesIfAvailable did a check to fetch the appcast feed from your server hence being inefficient.

> I wanted to implement a custom SPUUserDriver but I couldn't find a good example for this, and it seems that I need to implement all the methods declared on SPUUserDriver which could be overwhelming. Do we have a SPUUserDriverAdapter? I mean a class that has all the methods implemented and for what I just have to overwrite a specific method like showReadyToInstallAndRelaunch.

Technically you may be able to subclass `SPUStandardUserDriver` and override [-showUpdateFoundWithAppcastItem:state:reply:](https://sparkle-project.org/documentation/api-reference/Protocols/SPUUserDriver.html#/c:objc(pl)SPUUserDriver(im)showUpdateFoundWithAppcastItem:state:reply:) and [-showReadyToInstallAndRelaunch:](https://sparkle-project.org/documentation/api-reference/Protocols/SPUUserDriver.html#/c:objc(pl)SPUUserDriver(im)showReadyToInstallAndRelaunch:) but I don't strongly recommend it. There may be nuances with what the standard UI assumes and it may be fragile to rely on its implementation details in the future. You may by the way want to consider handling informational-only updates as a fail-safe.

A custom UI does provide some opportunity to streamline your updater experience. For example Sparkle transitions from two types of windows only because it has to show release notes / install options which aren't relevant for you. Simple examples are `SUPopUpTitlebarUserDriver` and `SPUCommandLineUserDriver` in the repository; `SPUStandardUserDriver` is a more complex one obviously.

For completeness sake there is also the option of enabling downloading/installing of updates automatically (`automaticallyDownloadsUpdates`), and there is [a delegate option](https://sparkle-project.org/documentation/api-reference/Protocols/SPUUpdaterDelegate.html#/c:objc(pl)SPUUpdaterDelegate(im)updater:willInstallUpdateOnQuit:immediateInstallationBlock:) to handle installing/relaunching an update that was automatically set to install. With this approach, users would not have to wait for an update to download and an update can be installed immediately. However it is not a complete approach because automatic installing cannot be done in some circumstances (notably when user needs to authorize installing an update when they don't have sufficient write permission, or for an info-only update). In those cases, it would fall back to the user driver showing the update. And if you don't want automatic downloading/installing, then this isn't for you.

**Reply by mbsanchez** [reply]
No, I'm not able to subclass `SPUStandardUserDriver` it seems that this class is private I can only subclass `SPUStandardUserDriverDelegate` and `SPUUserDriver`.

<img width="775" alt="image" src="https://github.com/sparkle-project/Sparkle/assets/143726/f83fabd9-7389-422a-b1e4-a42b4874b88e">

**Reply by zorgiepoo** [reply]
That's not subclassing. You're confusing protocols and classes. This is obj-c fundamentals.

**Reply by mbsanchez** [reply]
you might be right, I'm not an obj-c developer, I'm a C++ developer. I'm using this obj-c code in a Qt application.

**Reply by zorgiepoo** [reply, marked as answer]
If you're working with objc you should pick up the fundamentals and not guess at things.

This is a short subclassing sample:

```objc
@interface SUAppUserDriver : SPUStandardUserDriver
@end

@implementation SUAppUserDriver
{
    void (^_installUpdateBlock)(SPUUserUpdateChoice);
}

- (void)showUpdateFoundWithAppcastItem:(SUAppcastItem *)appcastItem state:(SPUUserUpdateState *)state reply:(void (^)(SPUUserUpdateChoice))reply
{
    _installUpdateBlock = [reply copy];
    
    // Show floating window prompt. When user clicks install invoke _installUpdateBlock(SPUUserUpdateChoiceInstall)
    // If appcastItem.informationOnlyUpdate, reply SPUUserUpdateChoiceDismiss and show user the appcastItem.infoURL instead
}

- (void)showReadyToInstallAndRelaunch:(void (^)(SPUUserUpdateChoice))reply
{
    reply(SPUUserUpdateChoiceInstall);
}

@end
```

Again this is not recommended compared to fully using your own UI.

**Reply by mbsanchez** [reply]
Sorry to bother you, but I've already fixed the code, and now, the error is gone. However, I'm wondering if the doc includes a full use case of how to extend the `SPUStandardUserDriver` as now, I can find the `setUserDriver` method, it is not at the updater object or updaterDeletegate either.

**Comment by mbsanchez** [comment]
I found the way to use the custom UserDriver, but I get not notified when an update is found. My code is like this:
```
    self.serDriver = [[AppUserDriver alloc] initWithHostBundle:[NSBundle mainBundle] delegate:nil];
    self.updater = [[SPUUpdater alloc] initWithHostBundle:[NSBundle mainBundle] applicationBundle:[NSBundle mainBundle] userDriver:self.userDriver delegate:Nil];

    [self.updater startUpdater:(nil)];
    [self.updater checkForUpdates];
```

What could be happening ?

**Reply by zorgiepoo** [reply]
Your user driver needs to show the update. That code is on you.. I don't know what your AppUserDriver implementation is. Also you should check startUpdater: actually succeeds (check return value). Also you are using the wrong API for checking updates. You want checkForUpdatesInBackground. Read the API documentation.

You can insert logs in your user driver methods to see if they are reached at least.

**Reply by mbsanchez** [reply]
Thank you so much @zorgiepoo I figured out what was happening, I just forget to remove the call to `checkForUpdateInformation`. Now it is working correctly. But I've got another question and I don’t know if I should open another thread for that. I just wondering If my application is double-signed (Sparkle and DeveloperId), even if I rotate the Sparkle signature it should accept the update, but it is telling me that the signature is **incorrect**.

**Reply by mbsanchez** [reply]
The errors I'm getting are:

> info 19:05:17.333976+0200	MyApp	Extracting using '/usr/bin/ditto' '-x' '-k' '-' < '/Users/mbsanchez/Library/Caches/com.MyApp.UI/org.sparkle-project.Sparkle/PersistentDownloads/twygSTXSN/MyApp 2.1.10/MyApp-2.1.10.pkg.zip' '/Users/mbsanchez/Library/Caches/com.MyApp.UI/org.sparkle-project.Sparkle/PersistentDownloads/twygSTXSN/MyApp 2.1.10'
error	19:05:17.424144+0200	MyApp	There is no DSA signature to check
error	19:05:17.424179+0200	MyApp	DSA signature validation of the package failed. The update contains an installer package, and valid DSA signatures are mandatory for all installer packages. The update will be rejected. Sign the installer with a valid DSA key or use an .app bundle update instead.

The AppCast items are like:
```
    <item>
      <title>Version 2.1.10</title>
      <sparkle:version>2.1.10</sparkle:version>
      <description>
      <![CDATA[
            <ul>
              <li>Stable version 2.1.10.</li>
            </ul>
          ]]>
      </description>
      <pubDate>Lun 15 mai 2023 21:00:49 CEST</pubDate>
      <enclosure url="https://zesty-chaja-28857e.netlify.app/mac/MyApp-2.1.10.pkg.zip" type="application/octet-stream" sparkle:edSignature="Cm5CTAXe1R8usn/MsD4z+3kYhHtAaYdYAW3vOXJb/d7clVoWXIUhSCsn+F3IhS74ACKfstVotOOm0veAp/sZBQ==" length="51021246"/>
    </item>
    <item>
      <title>Version 2.1.8</title>
      <description>
      <![CDATA[
            <ul>
              <li>Stable version 2.1.8.</li>
            </ul>
          ]]>
      </description>
      <pubDate>Ven  5 mai 2023 18:49:40 CEST</pubDate>
      <enclosure url="https://zesty-chaja-28857e.netlify.app/mac/MyApp-2.1.8.pkg.zip" sparkle:version="2.1.8" length="51665939" type="application/octet-stream" sparkle:dsaSignature="MC0CFD1iIsICqmAbkKFOs+R5mLuzkD8CAhUAhZM+fqI2QVIol0k3iAx7X0a4swg="/>
    </item>
```
BTW: I'm a package installer update

**Reply by mbsanchez** [reply]
Sorry, it seems that I need to migrate in two steps. It is working perfectly now with a double signature. However why it is required to have `sparkle:installationType="package"` it was not required in version 1.20.0

**Reply by zorgiepoo** [reply]
Yeah Apple developer ID code signing is not checked if you're using a package installer. The installationType became required due to Sparkle 2 re-architecturing the authorization code and isolating the installer helper. Sparkle's installer helper runs as a separate process and is responsible for extracting, verifying, and installing the update. It can run privileged (as root) when needed, which is always the case for package updates. If you distribute a zip file, Sparkle needs to know if the archive contains a package or is a regular app bundle update to properly determine if the installer helper, which extracts it, should be run as root.

It is not needed if you switch to [bare packages](https://sparkle-project.org/documentation/package-updates/) in the future because you wouldn't be putting the package inside an archive. Of course, developers should also avoid shipping pkg's for their apps unless they have a really good reason (e.g, legacy kexts)..

**Reply by mbsanchez** [reply]
Thank you so much @zorgiepoo, and yes, I have a good reason to distribute my app as a pkg installer. I hope you don't deprecate this in the future. :)
